src/lib.rs

```diff
@@ -1,29 +1,37 @@
+pub mod args;
+pub mod commands;
+pub mod error;
+
 use std::{ffi::OsString, path::Path};
 
-use commands::{nsec3hash::Nsec3Hash, LdnsCommand};
+use clap::Parser;
 
-pub use self::args::Args;
+use args::Args;
+use commands::{nsec3hash::Nsec3Hash, LdnsCommand};
+use error::Error;
 
-pub mod args;
-pub mod commands;
-pub mod error;
+pub fn parse_args<I: IntoIterator<Item = OsString>, T: Fn() -> I>(
+    args_provider: T,
+) -> Result<Args, Error> {
+    try_ldns_compatibility(args_provider()).or_else(|_| {
+        Args::try_parse_from(args_provider()).map_err(|err| Error::new(err.to_string().as_str()))
+    })
```

Comment:
This is subtly wrong. The `try_ldns_compatibility` function has three outcomes:
1. The command was not matched
2. The command matched and parsing succeeded
3. The command matched and parsing failed

You've now combined the first and third case even though they should be handled differently.

I think you're also removing the color from the clap output by turning it into a string.